clang/include/clang/CAS/CASOptions.h

```diff
@@ -67,17 +67,33 @@ class CASOptions : public CASConfiguration {
   /// Get a CAS defined by the options above. Future calls will return the same
   /// CAS instance... unless the configuration has changed, in which case a new
   /// one will be created.
+  ///
+  /// If \p CreateEmptyCASOnFailure, returns an empty in-memory CAS on failure.
+  /// Else, returns \c nullptr on failure.
   std::shared_ptr<llvm::cas::CASDB>
   getOrCreateCAS(DiagnosticsEngine &Diags,
                  bool CreateEmptyCASOnFailure = false) const;
 
+  /// Get a CAS defined by the options above. Future calls will return the same
+  /// CAS instance, even if the configuration changes again later.
+  ///
+  /// The configuration will be wiped out to prevent it being observable or
+  /// affecting the output of something that takes \a CASOptions as an input.
+  /// This also "locks in" the return value of \a getOrCreateCAS(): future
+  /// calls will not check if the configuration has changed.
+  std::shared_ptr<llvm::cas::CASDB>
+  getOrCreateCASAndHideConfig(DiagnosticsEngine &Diags);
```

Comment:
Can you say a bit about why you decided to do this at CAS creation time? Naively, I would have expected to copy and canonicalize the options when creating the cache key, or else save-and-restore to avoid the copy for performance. I don't love the idea of permanently clearing the options, since it means examining the invocation gives the wrong answer.  Do we have any other invocation options that behave this way?

Reply:
I did it this way because `CASOptions` is used in a few different places to get access to the CAS. It's important that none of these emits a diagnostic or uses the path in some way that affects output.

IMO, the logic should be:
1. Extract bits of CompilerInvocation specific to this compilation that shouldn't be part of the cache key.
    - For now, that's the CAS configuration.
    - Eventually, also diagnostics output configuration (canonicalize to serialize!)
    - Eventually, diagnostics that are handled externally (like `-Rcompile-job-cache*`).
    - Eventually, maybe, could/should prefix-mapping go here instead of in the scanner?
3. Canonicalize CompilerInvocation (removing everything extracted from (1))
4. Create a cache key, based on the canonicalized CompilerInvocation
5. Compute the result.
    - Return the cached result, if any.
    - Else, run a compile job action with the canonicalized CompilerInvocation.
    - Pass in (a) the CAS itself and (b) hooks to capture live-streamed partial results for anything that should be live-streamed for *this* compilation (e.g., diagnostics).
    - Save the result in the cache.
6. "Replay" the result.
    - De-canonicalize the result and output it as appropriate.
    - If the action was just run and something was live-streamed, no need to repeat.

> Do we have any other invocation options that behave this way?
I doubt it. There isn't anything else trying to do principled information hiding like this (the other caching is raw-lex caching; it hides information by ONLY taking `LangOptions` and nothing else). State is rarely stored in CompilerInvocation at all (it's odd to have a `CASDB` on there), but that seems like the "best" way to feed `getVFSFromCompilerInvocation()`.

Reply:
Okay, thanks for explaining.

Reply:
I documented this in 3a65c038937aab559fe13411fc7bc69e0344c0d6. Going to merge right away (without waiting for review) but feel free to send comments if you have them.